File: Java/solution28.java


public class solution28 {
    public int strStr(String haystack, String needle) {
        if (needle == null || needle.length() < 1) {
            return 0;
        }
        return kmpMatchFunction(haystack, needle);
    }

    public int kmpMatchFunction(String target, String pattern) {
        if (pattern == null || target == null) {
            return -1;
        }
        char[] t = target.toCharArray(), p = pattern.toCharArray();
        int[] lps = buildLps(p);
        int x = 0, y = 0;
        while (x < t.length) {
            if (t[x] == p[y]) {
                x++;
                y++;
            } else if (y != 0) {
                y = lps[y - 1];
            } else {
                x++;
            }
            if (y == p.length) {
                return x - y;
            }
        }
        return -1;
    }

    public int[] buildLps(char[] t) {
        int[] res = new int[t.length];
        res[0] = 0;
        int length = 0;
        for (int i = 1; i < res.length; i++) {
            while (length > 0 && t[i] != t[length]) {
                length = res[length - 1];
            }
            if (t[length] == t[i]) {
                length++;
            }
            res[i] = length;
        }
        return res;
    }
}

File: Cxx/solution28.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int strStr(string haystack, string needle) {
    	if (needle.size() == 0) {
    		return 0;
    	}
    	vector<int> f = preBuild(needle);
    	for (int i = 0, j = 0; i < haystack.size(); i++) {
    		while (j > 0 && haystack[i] != needle[j]) {
    			j = f[j - 1];
    		}
    		if (haystack[i] == needle[j]) {
    			j++;
    		}
    		if (j == needle.size()) {
    			return i - j + 1;
    		}
    	}
    	return -1;
    }

    vector<int> preBuild(string	input) {
    	vector<int> res = vector<int>(input.size(), 0);
    	for (int i = 1, j = 0; i < res.size(); i++) {
    		while (j > 0 && input[j] != input[i]) {
    			j = res[j - 1];
    		}
    		if (input[j] == input[i]) {
    			j++;
    		}
    		res[i] = j;
    	}
    	return res;
    }
};